src/shims/intrinsics.rs

```diff
@@ -440,15 +441,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             "assert_uninit_valid" => {
                 let ty = substs.type_at(0);
                 let layout = this.layout_of(ty)?;
-                // Return here because we panicked instead of returning normally from the intrinsic.
+                // Abort here because the caller might not be panic safe.
                 if layout.abi.is_uninhabited() {
-                    return this.start_panic(&format!("attempted to instantiate uninhabited type `{}`", ty), unwind);
+                    Machine::abort(this, Some(format!("attempted to instantiate uninhabited type `{}`", ty)))?;
```

Comment:
You can just use `throw_machine_stop!` here, to avoid having to change `Machine::abort`.